File: app/src/main/java/fhaachen/snakegame/MainActivity.java
package fhaachen.snakegame;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GestureDetectorCompat;

import fhaachen.snakegame.ui.GameStage;

public class MainActivity extends AppCompatActivity implements SensorEventListener, GestureDetector.OnGestureListener {
    private GameStage gameStage;
    private SensorManager sensorManager;
    private Sensor accelerometer;
    private GestureDetectorCompat gestureDetector;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Create a game stage
        gameStage = new GameStage(this);

        // Set game stage as view
        setContentView(gameStage);

        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager != null) {
            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        } else {
            //Disable Tilt setting if accelerometer is not available
            TextView tiltSetting = findViewById(R.id.control_tilt_button);
            if (tiltSetting != null) {
                tiltSetting.setClickable(false);
            }
        }

        gestureDetector = new GestureDetectorCompat(this, this);
    }

    @Override
    public void onBackPressed() {
        gameStage.showPauseDialog();
    }

    @Override
    protected void onResume() {
        super.onResume();
        sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL);
        gameStage.resume();
    }

    @Override
    protected void onPause() {
        super.onPause();

        // Stop game stage thread
        gameStage.pause();
        sensorManager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        gameStage.onSensorChanged(event, accelerometer);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        //Do nothing
    }

    @Override
    public void onPointerCaptureChanged(boolean hasCapture) {
        //Do nothing
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return gestureDetector.onTouchEvent(event) || super.onTouchEvent(event);
    }

    @Override
    public boolean onDown(MotionEvent e) {
        return false;
    }

    @Override
    public void onShowPress(MotionEvent e) {
        //Do nothing special
    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        return false;
    }

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        return false;
    }

    @Override
    public void onLongPress(MotionEvent e) {
        //Do nothing special
    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        return gameStage.onFling(velocityX, velocityY);
    }

    public void onControlSelected(View v) {
        gameStage.onControlSelected(v);
    }

    public void onThemeSelected(View v) {
        gameStage.onThemeSelected(v);
    }
}

File: app/src/main/java/fhaachen/snakegame/MainActivity.kt
package fhaachen.snakegame

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GestureDetectorCompat
import fhaachen.snakegame.ui.GameStage

class MainActivity : AppCompatActivity(), SensorEventListener, GestureDetector.OnGestureListener {
    private lateinit var gameStage: GameStage
    private var sensorManager: SensorManager? = null
    private var accelerometer: Sensor? = null
    private lateinit var gestureDetector: GestureDetectorCompat

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Create a game stage
        gameStage = GameStage(this)
        // Set game stage as view
        setContentView(gameStage)

        //Sensors
        sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
        if (sensorManager != null) {
            accelerometer = sensorManager!!.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        } else {
            //Disable Tilt setting if accelerometer is not available
            val tiltSetting = findViewById<TextView>(R.id.control_tilt_button)
            if (tiltSetting != null) {
                tiltSetting.isClickable = false
            }
        }
        gestureDetector = GestureDetectorCompat(this, this)
    }

    override fun onBackPressed() {
        gameStage.showMenuDialog()
    }

    override fun onResume() {
        super.onResume()
        if (accelerometer != null) {
            sensorManager!!.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL)
        }
        gameStage.resume()
    }

    override fun onPause() {
        super.onPause()
        // Stop game stage thread
        gameStage.pause()
        sensorManager!!.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        gameStage.onSensorChanged(event, accelerometer!!)
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        //Do nothing
    }

    override fun onPointerCaptureChanged(hasCapture: Boolean) {
        //Do nothing
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return gestureDetector.onTouchEvent(event) || super.onTouchEvent(event)
    }

    override fun onDown(e: MotionEvent): Boolean {
        return false
    }

    override fun onShowPress(e: MotionEvent) {
        //Do nothing special
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean {
        return false
    }

    override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
        return false
    }

    override fun onLongPress(e: MotionEvent) {
        //Do nothing special
    }

    override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
        return gameStage.onFling(velocityX, velocityY)
    }

    fun onControlSelected(v: View) {
        gameStage.onControlSelected(v)
    }

    fun onThemeSelected(v: View) {
        gameStage.onThemeSelected(v)
    }
}